Mathlib/Topology/Category/Profinite/Nobeling.lean

```diff
@@ -0,0 +1,1556 @@
+/-
+Copyright (c) 2023 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.Algebra.Category.ModuleCat.Free
+import Mathlib.Topology.Category.Profinite.CofilteredLimit
+import Mathlib.Topology.Category.Profinite.Product
+import Mathlib.Topology.LocallyConstant.Algebra
+
+/-!
+
+# Nöbeling's theorem
+
+This file proves Nöbeling's theorem.
+
+## Main result
+
+- `Nobeling`: For `S : Profinite`, the `ℤ`-module `LocallyConstant S ℤ` is free.
+
+## Proof idea
+
+We follow the proof of theorem 5.4 in [scholze2019condensed], in which the idea is to embed `S` in
+a product of `I` copies of `Bool` for some sufficiently large `I`, and then to choose a
+well-ordering on `I` and use ordinal induction over that well-order. Here we can let `I` be
+the set of clopen subsets of `S` since `S` is totally separated.
+
+For `i : I`, let `e S i : LocallyConstant (S : Set (I → Bool)) ℤ` denote the map
+`fun f ↦ (if f.val i then 1 else 0)`.
+
+The basis will consist of products `e S iᵣ * ⋯ * e S i₁` with `iᵣ > ⋯ > i₁` which cannot be written
+as linear combinations of lexicographically smaller products.
+
+What is proved by ordinal induction is that this set is linearly independent. The fact that it
+spans can be proved directly.
+
+## References
+
+- [scholze2019condensed]: *Lectures on Condensed Mathematics*, 2019.
+- [scholze2019condensed], Theorem 5.4.
+-/
+
+universe u
+
+namespace NobelingProof
+
+variable {I : Type u} [Inhabited I] [LinearOrder I] [IsWellOrder I (·<·)] (C : Set (I → Bool))
+
+open Profinite ContinuousMap CategoryTheory Limits Opposite Submodule
+
+section Projections
+
+variable (J K L : I → Prop) [∀ i, Decidable (J i)] [∀ i, Decidable (K i)] [∀ i, Decidable (L i)]
+
+/-- The projection mapping everything that satisfies `J i` to itself, and everything else to
+    `false` -/
+def Proj : (I → Bool) → (I → Bool) :=
+  fun c i ↦ if J i then c i else false
+
+@[simp]
+theorem continuous_proj :
+    Continuous (Proj J : (I → Bool) → (I → Bool)) := by
+  dsimp [Proj]
+  apply continuous_pi
+  intro i
+  split
+  · apply continuous_apply
+  · apply continuous_const
+
+/-- The image of `Proj π J` -/
+def π : Set (I → Bool) := (Proj J) '' C
```

Comment:
This definition makes me sad. Why is the order of the inputs `C` then `J`?
Consider the alternative `π '` (you can run this code just after the definition of `π`):
```lean
variable (C : Set (I → Bool)) (J : I → Prop)
  [(i : I) → Decidable (J i)]

#check π C J -- Set (I → Bool)

def π' {I : Type u} (J : I → Prop) [(i : I) → Decidable (J i)]
    (C : Set (I → Bool)) : Set (I → Bool) := Set.image (Proj J) C

example : π' J C = π C J := rfl

example : π' J = Set.image (Proj J) := rfl
```
`π'` and `π` are "the same thing" as far as a mathematician is concerned, but if you switch the order of the inputs then you can see that we already have `\pi'` (it's `Set.image \circ Proj`) so we don't need another definition at all. Is it possible to get a machine to do the refactor replacing all `\pi` by `\pi'` and then you can just make `\pi'` an `abbrev` and then rename it to `pi`. This will make automation work better and might also help getting a better understanding of this `\pi` map.

Reply:
OK from talking to Mario it sounds like this is hard to do programatically and arguably pointless because it's a one-off definition in a self-contained proof.